packages/headless/src/controllers/insight/user-actions/headless-user-actions.ts

```diff
@@ -0,0 +1,145 @@
+import {UserActionTimeline} from '../../../api/service/insight/user-actions/user-actions-response';
+import {InsightEngine} from '../../../app/insight-engine/insight-engine';
+import {configuration, insightUserActions} from '../../../app/reducers';
+import {
+  fetchUserActions,
+  incrementNumberOfSessionsAfter,
+  incrementNumberOfSessionsBefore,
+  registerUserActions,
+} from '../../../features/insight-user-actions/insight-user-actions-actions';
+import {InsightAPIErrorStatusResponse} from '../../../insight.index';
+import {
+  ConfigurationSection,
+  InsightUserActionSection,
+} from '../../../state/state-sections';
+import {loadReducerError} from '../../../utils/errors';
+import {
+  buildController,
+  Controller,
+} from '../../controller/headless-controller';
+
+export interface UserActionsProps {
+  /**
+   * The options for the `UserActions` controller.
+   */
+  options: UserActionsOptions;
+}
+
+export interface UserActionsOptions {
+  /**
+   * The ticket creation date in the [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) format.
+   */
+  ticketCreationDate: string;
+  /**
+   * The number of sessions to include occuring before the ticket creation session.
+   * @defaultValue `0`
+   */
+  numberSessionsBefore?: number;
+  /**
+   * The number of sessions to include occuring after the ticket creation session.
+   * @defaultValue `0`
+   */
+  numberSessionsAfter?: number;
+  /**
+   * The names of custom events to exclude.
+   * @defaultValue `[]`
+   */
+  excludedCustomActions?: string[];
+}
+
+export interface UserActionsState {
+  /**
+   * The timeline of user actions.
+   */
+  timeline: UserActionTimeline;
+  /**
+   * The number of sessions included before the ticket creation date.
+   */
+  numberSessionsBefore: number;
+  /**
+   * The number of sessions included after the ticket creation date.
+   */
+  numberSessionsAfter: number;
+  /**
+   * The names of custom actions to exclude from the user actions.
+   */
+  excludedCustomActions: string[];
+  /**
+   * The ticket creation date in the [ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) format.
+   */
+  ticketCreationDate?: string;
+  /**
+   * `true` if fetching the user actions is in progress and `false` otherwise.
+   */
+  loading: boolean;
+  /**
+   * The error response if fetching the user actions failed.
+   */
+  error?: InsightAPIErrorStatusResponse;
+}
+
+/**
+ * The UserActions controller is responsible for fetching user actions surrounding a given case creation event.
+ */
+export interface UserActions extends Controller {
+  /**
+   * Fetch the list of user actions surrounding the ticket creation.
+   */
+  fetchUserActions(): void;
+  /**
+   * Fetch the list of user actions including one more session preceding the ticket creation.
+   */
+  fetchWithPrecedingSession(): void;
```

Comment:
Out of curiosity, do we currently plan to have a button in the front-end that allows to fetch one more preceding/following session?
I dont see any in the mockups so I'm assuming this action is made for potential future use right?

Reply:
It's right there in the mockup. 😅 "Show preceding sessions", "Show following sessions". Hmmm I think i see a pattern here with your previous question. Maybe we interpreted it differently.
Did you see us requesting all sessions at once and putting them in the state, then showing/hiding them in the UI?
I was treating it as multiple requests.
Both would work.

If we fetch them all at once then we would set a default number of sessions before/after on load and that's that. But I'm not sure how we determine the best setting.

If we fetch new sessions based on user clicks then it's more requests but potentially lighter payloads and we don't have to guess.

Reply:
Ok so I seem to have I misunderstood the UX design. What is prescribed by UX is to fetch 3 following sessions and 2 preceding sessions on load. That would be "All the timeline sessions". Then the UI can show/hide the sessions but it will already have all of them.
This sounds good to me so I would remove these two actions. Sorry for the confusion that's my bad.

Reply:
Yes that's exactly what I thought, I thought that we are requesting all the sessions all at once.
Like you mentioned both would work, however I have some considerations I want to mention:
- The "Show preceding sessions" button is supposed to directly show **2** preceding  sessions and not display the preceding sessions one by one.
- The "Show following sessions" button is supposed to directly show **3** following  sessions and not display the following sessions one by one. (That's why I understood that we will fetch all the sessions at once, cause we know exactly how many sessions we wanna display). 
- If we go with second approach of fetching new sessions based on user clicks, wouldn't be better to consider having a loading state for every type of user sessions,  preceding, following and case creation session? Because as a user of the QuanticUserActionTimeline, when I click for example on  the "Show preceding sessions" button, it would feel a bit weird to see all the component loading again, I would expect to only the preceding sessions to be loading.

Reply:
Yeah your points are all super valid which is why i'm happy to do exactly what's in the mockups. We do one fetch of (main session + 2 before + 3 after). If it has fewer sessions it just has fewer sessions. Then in the quantic component one click shows all the preceding or all the following sessions with no load time. Easy, simple.